# APS106 - Fundamentals of Computer Programming
## Week 11 | Lecture 1 (11.2) - More files, file methods, CSV module, creating your own modules

### This Week
| Lecture | Topics |
| --- | --- |
| 12.1 | Object-Oriented Examples |
| **12.1** | **More files, file methods, CSV module, creating your own modules** |
| 12.3 | Cancelled |

### Lecture Structure
1. [Writing a dictionary to file](#section1)
2. [More advanced file reading](#section2)
3. [The with statement](#section3)
4. [CSV files](#section4)
5. [Modules](#section5)

 <a id='section1'></a>
 
## 1. Writing a dictionary to file

In [3]:
students = {'Kendrick': 'A+', 'Dre': 'C-', 'Snoop': 'B'} 

# create a file
myfile = open("grades.txt", "w")

# store dictionary items to the file
for student in students:
    myfile.write(student + ',' + students[student] + '\n')

# close the file    
myfile.close()

<a id='section2'></a>
 
## 2. More advanced file reading

### for line in file

In [5]:
#declaring filename to change easily if you want to try with different files
filename = 'grades.txt'

In [6]:
# Approach: for line in file
# When to use it: When you want to process the file line-by-line
# Example code

myfile = open(filename, 'r')
contents = ''
for line in myfile: # each time through the loop line contains one line of the file
    contents += line
    #print(line)  #why is there a gap between rows?
myfile.close()

print(contents)
# by the end of this contents contains the entire contets of the file

Kendrick,A+
Dre,C-
Snoop,B



### readlines

In [37]:
# Approach: readlines
# When to use it: When you want to process the file line-by-line with an index
# Example code
myfile = open(filename, 'r')
lines = myfile.readlines() # lines is a list of strings. Each entry in lines is a line of the file
myfile.close()
print(lines)

['Kendrick,A+\n', 'Dre,C-\n', 'Snoop,B\n']


In [12]:

students = {}
myfile = open("grades.txt", "r")

# read each line of the file
for line in myfile:
    # find indices for slicing each line
    ind1 = line.find(',')
    ind2 = line.find('\\')
    name = line[:ind1]
    grade = line[ind1+1:ind2]
    students[name] = grade

myfile.close()

print(students)

{'Kendrick': 'A+', 'Dre': 'C-', 'Snoop': 'B'}


<a id='section3'></a>
## 3. The with Statement

Notice that whenever we open a file, we need to be careful to close it again. Python provides a nice way to open and then automatically close a file using a `with` block.

```
with open(«filename», «mode») as «variable»:
      «body»
```

The file is opened at the beginning and **automatically closed** at the end of the body. 


In [48]:
def f(file):
    print(file.read())

with open('test.txt', 'r') as file:  #test.txt is from beginning of notebook
    f(file)
    
print("The next line")

CATS!
I <3 my second sentence

The next line


In [ ]:
with open("flanders.txt", 'w') as flanders_file:
    
    content = flanders_file.()
    
    for line in flanders_file:
        print(line, end="")
print('hello')

 <a id='section4'></a>
## 4. CSV Files

The CSV format (comma separated values) is very commonly used to represent the data in a spreadsheet. 

For example a spreadsheet such as:

Name|Test1|Test2|Final
----|-----|-----|-----
Kendrick|100|50|29
Dre|76|32|33
Snoop|25|75|95

is represented as a file like this:

```
Name,Test1,Test2,Final
Kendrick,100,50,29
Lamar,76,32,33
Snoop,25,75,95
```

We can, of course, access this files using the techniques above.

In [8]:
csv_file = open('grades.csv','w')
csv_file.write('''Name,Test1,Test2,Final
Kendrick,100,50,29
Dre,76,32,33
Snoop,25,75,95
''')
csv_file.close()

In [22]:
with open('grades.csv','w') as csv_file:
    csv_file.write('Name,Test1,Test2,Final\nKendrick,100,50,29\nDre,76,32,33\nSnoop,25,75,95')


In [11]:
with open('grades.csv', 'r') as file:
    for line in file:
        print(line, end='')

Name,Test1,Test2,Final
Kendrick,100,50,29
Dre,76,32,33
Snoop,25,75,95

Notice that you have the information about each row and also the commas. If you are going to process this data, you are going to need to **parse** it. That means , for example, to discard the commas (as they just separate the data and are not otherwise meaningful), to extract the integers from the string.

One of the great things about Python is the existence of many modules that give us the ability to easily do many things, like reading and writing CSV files. 

Reading of CSV files can be done using the CSV reader. You can construct a reader object using `csv.reader()` which takes the file object as input. The reader object can be used to iterate through the contents of the CSV file, similarly to how a file object was used to iterate through the contents in a text file.

The difference between the two is that the file method `read(`) returns the entire contents of the file as one long string, whereas, the CSV `reader()` returns an object which can be iterated through. The reader object holds each row as a list of strings and can be iterated through row by row. 

Example: Read each row of a CSV file

In [23]:
import csv

with open('grades.csv', 'r') as file:
    print(file)
    grades_reader = csv.reader(file) # create csv.reader object with an open file
    print(grades_reader) #what does grades_reader look like? just some object, but it is iterable!
    row_num = 1
    for row in grades_reader:           # the cvs.reader is an iterable!
        print(row)
        #print('Row #', row_num, ':', row)
        row_num += 1

<_io.TextIOWrapper name='grades.csv' mode='r' encoding='UTF-8'>
['Name', 'Test1', 'Test2', 'Final']
['Kendrick', '100', '50', '29']
['Dre', '76', '32', '33']
['Snoop', '25', '75', '95']


If we didn’t have a CSV file created, we could create one by:
- creating a CSV writer object
- using the writerow() method to populate it with data

Example: In the previous grade example there were a few marking errors on the final exam. 


In [19]:
import csv

rows = [['Name', 'Test1', 'Test2', 'Final'],
        ['Kendrick', '100', '50', '69'],
        ['Dre', '76', '32', '53'],
        ['Snoop', '25', '75', '100']]

with open('grades_new.csv', 'w') as csvfile:
    print(csvfile)
    grades_writer = csv.writer(csvfile)
    print(grades_writer)

    for row in rows:
        grades_writer.writerow(row)
        

<_io.TextIOWrapper name='grades_new.csv' mode='w' encoding='UTF-8'>
Help on built-in function writerow:

writerow(...) method of _csv.writer instance
    writerow(iterable)
    
    Construct and write a CSV record from an iterable of fields.  Non-string
    elements will be converted to string.



<a id='section5'></a>
## 5. Modules
### What is a Module?
A module in Python is a file containing Python definitions and statements. A module can define functions, classes, and variables, and it can also include executable code. You can use a module in your code by importing it. Python comes with many built-in modules, such as the `math`, `random`, `csv`, and `turtle` modules that we have learned about thus far in the course.

### Importing Modules
To use a module in your code, you need to import it. The syntax for importing a module is:

```python
import module_name
```
For example, let's import the `math` module.

In [2]:
import math

Once you have imported a module, you can use its functions and variables by prefixing them with the module name. For example, to use the `pi` constant from the `math` module, you would use:

In [3]:
print(math.pi)

3.141592653589793


Or, we can use functions to calculate the sine of an angle.

In [4]:
print(math.sin(math.radians(90)))

1.0


#### from X import Y Syntax
Sometimes, you might only want to import specific functions or variables from a module, rather than importing the entire module. In this case, you can use the `from X import Y` syntax. The syntax for using this syntax is:
```python
from module_name import variable_name
from module_name import function_name
from module_name import class_name
```
For example, to import only the `pi` constant from the `math` module, you would use:

In [5]:
from math import pi

print(pi)

3.141592653589793


This creates a variable named `pi` in the global space with the value `3.141592653589793` of type `float`.

In [6]:
type(pi)

float

You can also import multiple functions or variables from a module by separating them with commas, like this:

In [7]:
from math import pi, sqrt

print(pi)
print(sqrt(4))

3.141592653589793
2.0


In this case, there is now a function definition called `sqrt` in the global space that is callable.

You'll notice that when you import functions, classes or variables in this way, we no longer need to access them using `math.`

#### from X import * Syntax
In Python, the from X import * syntax allows you to import all functions and variables from a module into your current namespace. This means that you can use the imported functions and variables without prefixing them with the module name (`math.`). For example, if you import all functions and variables from the math module, you could use `pi` directly, like this:

In [ ]:
from math import *

print(pi)

However, using the from `X import * syntax` is generally not recommended for several reasons.

1. It can make your code harder to read and understand. When you use the `from X import *` syntax, it’s not always clear where a function or variable is coming from. This can make it difficult to trace the source of bugs and errors in your code.

2. It can lead to naming conflicts. If you import all functions and variables from two different modules that have the same name, you’ll end up with naming conflicts. For example, if you import all functions and variables from both the `math` module and the `statistics` module, you’ll end up with two functions named `fabs()`. This can cause confusion and errors in your code.

3. It can make your code less efficient. When you use the `from X import *` syntax, Python has to load and compile all the functions and variables from the module, even if you’re only going to use a few of them. This can slow down the loading time of your program and use up more memory than necessary.

In general, it’s better to be explicit about which functions and variables you’re importing from a module. This makes your code easier to read and understand, reduces the risk of naming conflicts, and can make your code more efficient. If you need to import a large number of functions and variables from a module, it’s better to import them explicitly, like this:

In [ ]:
from math import pi, sin, cos, tan

This way, you can be sure which functions and variables you’re using and avoid naming conflicts.

#### import X as Syntax
The default name given to modules in your code is the name of the module. For example, `math`, `random`, etc.

However, you can override the default module name using the `as` keyword. This is useful if the name of a module conflicts with a variable you have declared in your code. It is also useful if you want to reference another module that shares the same name.

Suppose we want to import the `turtle` module into your code but you already have a function called `turtle`.

In [9]:
turtle = 'Stewart'

import turtle

print(turtle)

<module 'turtle' from '/Users/benkinsella/anaconda3/lib/python3.10/turtle.py'>


Not I have overwritten the variable `turtle` with the module `tutle`. So avoid this error, I would use an alias when importing the `turtle` module. The syntax for importing a module is an alias is:

```python
import module_name as alias_name
```

In [10]:
turtle = 'Stewart'

import turtle as t

print(turtle)

Stewart


Now, when I want to use the turtle module, I can use the alias to access the functions, classes, and variables in the `turtle` module.

In [11]:
new_turtle = t.Turtle()  #running this may cause your Jupyter kernel to crash...

### How to Create Your Own Module in Python
Creating your own module is useful when the variables, functions, or classes you're creating in a `.py` file or `.ipynb` file are generic and likely to be used across many notebooks, projects, assignments, etc. For example, it would be silly to continually create `sine` functions in everythin notebook. This is why someone created the `math` module.

A module in Python is simply a file with a `.py` extension that contains Python code. To create a module, you just need to write your code in a `.py` file and save it with an appropriate name. When you import a module in Python, the interpreter searches for the module in the following locations:

- The current directory
- The directories listed in the PYTHONPATH environment variable
- The installation-dependent default directory

To ensure Python is able to file our modules, we will be placing our `.py` files in the same folder as this `ipynb` file. 

```bash
├── lectures
│   ├── week1
│   ├── week2
│   ├── week3
│   ├── week4
│   ├── week5
│   ├── week6
│   ├── week7
│   ├── week8
│   ├── week9
│   ├── week10
│   ├── week11
│   |   ├── lecture1
│   |   ├── lecture2
│   |   |   ├── notebooks
│   |   |   |   ├── week11_lecture1.ipynb
│   |   |   |   ├── geometry_calculations.py
│   |   |   ├── slides
│   |   ├── lecture3
│   ├── week12
│   ├── week13
```
In you're using `JupyterHub`, you can see that `geometry_calculations.py` is placed in the same folder as `week11_lecture1.ipynb`. If you're downloading these files from Quercus, please place them in the same folder.

Here is the code we have in `geometry_calculations.py`>

```python
import math


def rectangle_area(length, width):
    """
    (number, number) -> number
    Calculate the area of a rectangle.
    """
    return length * width


def circle_area(radius):
    """
    (number) -> number
    Calculate the area of a circle.
    """
    return math.pi * radius**2


def cylinder_volume(radius, height):
    """
    (number, number) -> number
    Calculate the volume of a cylinder.
    """
    return math.pi * radius**2 * height


def rectangular_prism_volume(length, width, height):
    """
    (number, numner, number) -> number
    Calculate the volume of a rectangular prism (3D rectangle).
    """
    return length * width * height


def cylinder_surface_area(radius, height):
    """
    (number, number) -> number
    Calculate the surface area of a cylinder.
    """
    return 2 * math.pi * radius * (radius + height)


def rectangular_prism_surface_area(length, width, height):
    """
    (number, number, number) -> number
    Calculate the surface area of a rectangular prism (3D rectangle).
    """
    return 2 * (length*width + length*height + width*height)
```

We have six functions and we have also imported the math module to access the `pi` constant.

Now, you can use this module in your other Python programs by importing it. We can import the circle and use its functionality:

In [1]:
import geometry_calculations

radius = 5
area = geometry_calculations.circle_area(radius)

print('The area of a circle with radius ', radius, ' is ', area, '.', sep='')

The area of a circle with radius 5 is 78.53981633974483.


You can see all the functions in `geometry_calculations.py` by clicking TAB below.

In [ ]:
geometry_calculations.  #<--- place your cursor after the . and press TAB.

Interestingly, because we imported the `math` module in `geometry_calculations.py`, we can access is as follows:

In [ ]:
geometry_calculations.math.pi

<a id='section6'></a>
## 6. Challenge: Build Your Own Module
Create a module called `basic_math`, with the following functions:

### Add: 1 + 2 = 3
```python
>>> print(add(1, 2))
3
```

### Subtract: 5 - 1 = 4
```python
>>> print(subtract(5, 1))
4
```

### Multiply: 2 * 3 = 6
```python
>>> print(multiply(2, 3))
6
```

### Divide: 6 / 12 = 0.5
```python
>>> print(divide(6, 12))
0.5
```
This function should return the following error if the user tries to divide by 0.
```python
>>> print(divide(6, 0))
Error. You cannot divide by zero.
```
Import your module.

In [ ]:
import basic_math

In [ ]:
print(basic_math.add(1, 2))

In [ ]:
print(basic_math.subtract(5, 1))

In [ ]:
print(basic_math.multiply(2, 3))

In [ ]:
print(basic_math.divide(6, 12))

In [ ]:
print(basic_math.divide(6, 0))